### PR TITLE
Signup api

### PR DESCRIPTION
**Pull request by mkrs2404**
Please review the PR

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -1,32 +1,24 @@
...
-	err = database.Ping()
-	handleError(err)
-
-	fmt.Println("Connected to Postgres")
-
-}
-
-func handleError(err error) {
+	err := godotenv.Load()
 	if err != nil {
```
Comment (jdk2588) [inline main.go:35]: No need for else block
```
if err != nil {
   log.Fatal("Error fetching")
}

<your logic here>
```
Read https://en.wikipedia.org/wiki/Guard_(computer_science)

File: .env [hunk 2]
```diff
@@ -0,0 +1,16 @@
+DB_HOST=localhost
```
Comment (jdk2588) [inline .env]: Create a separate `.env` file for test? Production credentials are never to be stored in source code.

File: api/controllers/signup_controller.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	c.JSON(http.StatusCreated, gin.H{
+		"access_key": token,
+	})
+}
+
+/*reportValidationFailure sends the response back with proper errors during validation.
+It returns whether there was a validation error or not*/
+func reportValidationFailure(err error, c *gin.Context) bool {
+	if err != nil {
```
Comment (jdk2588) [inline api/controllers/signup_controller.go]: if err == nil {
    return true
}

<do you code>
return false

File: api/controllers/signup_controller.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		})
+		c.Abort()
+		return
+	}
+
+	//Generating JWT token to send back as response
+	token, err := auth.GenerateToken(int(client.ID))
+	if err != nil {
+		c.JSON(http.StatusInternalServerError, gin.H{
+			"msg": "sign Up failed",
```
Comment (jdk2588) [inline api/controllers/signup_controller.go]: Start taking these error message as constants, which are repeated at multiple places
```
const SIGN_UP_FAILED="Sign up failed"
```

File: api/controllers/signup_controller.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		var validatorErr validator.ValidationErrors
+		var errorMsg string
+		//Checking the type of validation error
+		if errors.As(err, &validatorErr) {
+			for _, error := range validatorErr {
+				errorMsg += error.Field() + " : " + msgForTag(error.Tag()) + "; "
+			}
+			//Sending the error response
+			c.JSON(http.StatusBadRequest, gin.H{
+				"errorMsg": errorMsg[:len(errorMsg)-2],
```
Comment (jdk2588) [inline api/controllers/signup_controller.go]: What is the intent here? 

Can you add an example in comments section?

File: main.go [hunk 6]
```diff
@@ -1,32 +1,24 @@
...
-		log.Fatal(err)
+		log.Fatal("Error fetching the environment values")
+	} else {
+		database.Connect(os.Getenv("DB_HOST"), os.Getenv("DB_NAME"), os.Getenv("DB_USER"), os.Getenv("DB_PASSWORD"), os.Getenv("DB_PORT"), logger.Error)
```
Comment (jdk2588) [inline main.go]: Try reading from command line args. Why?
- Developers are forced to pass at runtime. Credentials will not be stored in repo.

**Review by jdk2588 — CHANGES_REQUESTED** [review]

File: api/controllers/signup_controller.go (hunk 5 above)
Comment (mkrs2404) [reply api/controllers/signup_controller.go]: If there are multiple validation failures, errorMsg will be something like this:
"errorMsg" : "Email : Invalid ; Plan : This field is required ;"
To remove the last 2 characters ( ;)  , I'm doing this

File: main.go (hunk 6 above)
Comment (mkrs2404) [reply main.go]: For test, should I keep the env file?

File: main.go (hunk 1 above)
Comment (mkrs2404) [reply main.go:35]: Resolved

File: api/controllers/signup_controller.go (hunk 3 above)
Comment (mkrs2404) [reply api/controllers/signup_controller.go]: Resolved

File: api/controllers/signup_controller.go (hunk 4 above)
Comment (mkrs2404) [reply api/controllers/signup_controller.go]: Resolved

**Pull request merged**